data/state/accountsDB.go

```diff
@@ -93,50 +93,46 @@ func (adb *AccountsDB) SaveAccount(account AccountHandler) error {
 		adb.journalize(entry)
 	}
 
-	baseAcc, ok := account.(baseAccountHandler)
-	if ok {
-		err = adb.saveCode(baseAcc, oldAccount)
-		if err != nil {
-			return err
-		}
-
-		err = adb.saveDataTrie(baseAcc)
-		if err != nil {
-			return err
-		}
+	err = adb.saveCodeAndDataTrie(oldAccount, account)
+	if err != nil {
+		return err
 	}
 
 	return adb.saveAccountToTrie(account)
 }
 
-func (adb *AccountsDB) getOldCodeHash(oldAcc AccountHandler) ([]byte, error) {
-	if check.IfNil(oldAcc) {
-		return nil, nil
+func (adb *AccountsDB) saveCodeAndDataTrie(oldAcc, newAcc AccountHandler) error {
+	baseNewAcc, newAccOk := newAcc.(baseAccountHandler)
+	baseOldAccount, _ := oldAcc.(baseAccountHandler)
+
+	if !newAccOk {
+		return nil
 	}
 
-	oldAccount, ok := oldAcc.(baseAccountHandler)
-	if !ok {
-		return nil, ErrWrongTypeAssertion
+	err := adb.saveDataTrie(baseNewAcc)
+	if err != nil {
+		return err
 	}
 
-	return oldAccount.GetCodeHash(), nil
+	return adb.saveCode(baseNewAcc, baseOldAccount)
 }
 
-func (adb *AccountsDB) saveCode(accountHandler baseAccountHandler, oldAcc AccountHandler) error {
+func (adb *AccountsDB) saveCode(newAcc, oldAcc baseAccountHandler) error {
 	// TODO when state splitting is implemented, check how the code should be copied in different shards
-	oldCodeHash, err := adb.getOldCodeHash(oldAcc)
-	if err != nil {
-		return err
+
+	var oldCodeHash []byte
+	if !check.IfNil(oldAcc) {
+		oldCodeHash = oldAcc.GetCodeHash()
 	}
 
-	newCode := accountHandler.GetCode()
+	newCode := newAcc.GetCode()
```

Comment:
newAcc is always not nil? If not, add a check above

Reply:
newAccount is always not nil. It is checked on L72, when SaveAccount() is called.